drivers/dac/dac_sam0.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2020 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam0_dac
+
+#include <errno.h>
+
+#include <drivers/dac.h>
+#include <soc.h>
+
+/*
+ * Maps between the DTS reference property names and register values
+ * TODO(#21273): replace once improved support for enum values lands.
+ */
+#define SAM0_DAC_REFSEL_0 DAC_CTRLB_REFSEL_INT1V_Val
+#define SAM0_DAC_REFSEL_1 DAC_CTRLB_REFSEL_AVCC_Val
+#define SAM0_DAC_REFSEL_2 DAC_CTRLB_REFSEL_VREFP_Val
+
+struct dac_sam0_cfg {
+	Dac *regs;
+	uint8_t pm_apbc_bit;
+	uint8_t gclk_clkctrl_id;
+	uint8_t refsel;
+};
+
+#define DEV_CFG(dev) ((const struct dac_sam0_cfg *const)(dev)->config_info)
+
+/* Write to the DAC. */
+static int dac_sam0_write_value(struct device *dev, uint8_t channel,
+				uint32_t value)
+{
+	const struct dac_sam0_cfg *const cfg = DEV_CFG(dev);
+	Dac *regs = cfg->regs;
+
+	regs->DATA.reg = (uint16_t)value;
+
+	return 0;
+}
+
+/*
+ * Setup the channel.  As the SAM0 has one fixed width channel, this validates
+ * the input and does nothing else.
+ */
+static int dac_sam0_channel_setup(struct device *dev,
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	if (channel_cfg->channel_id != 0) {
+		return -EINVAL;
+	}
+	if (channel_cfg->resolution != 10) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/* Initialise and enable the DAC. */
+static int dac_sam0_init(struct device *dev)
+{
+	const struct dac_sam0_cfg *const cfg = DEV_CFG(dev);
+	Dac *regs = cfg->regs;
+
+	/* Enable the GCLK */
+	GCLK->CLKCTRL.reg = cfg->gclk_clkctrl_id | GCLK_CLKCTRL_GEN_GCLK0 |
+			    GCLK_CLKCTRL_CLKEN;
+
+	/* Enable the clock in PM */
+	PM->APBCMASK.reg |= 1 << cfg->pm_apbc_bit;
+
+	/* Reset then configure the DAC */
+	regs->CTRLA.bit.SWRST = 1;
+	while (regs->STATUS.bit.SYNCBUSY) {
+	}
+
+	regs->CTRLB.bit.REFSEL = cfg->refsel;
+	regs->CTRLB.bit.EOEN = 1;
+
+	/* Enable */
+	regs->CTRLA.bit.ENABLE = 1;
```

Comment:
Should we also configure RUNSTBY in the devicetree?  It seems to me for power sensitive applications, this would be a useful configuration option.

Reply:
Same comment as above - doing this globally in a separate PR SGTM.

Reply:
Optimistically closing.